wgengine/wgcfg/parser.go

```diff
@@ -122,62 +118,70 @@ func FromUAPI(r io.Reader) (*Config, error) {
 	return cfg, nil
 }
 
-func (cfg *Config) handleDeviceLine(key, value string) error {
-	switch key {
-	case "private_key":
-		k, err := parseKeyHex(value)
-		if err != nil {
+func parseKeyHex(s []byte, dst []byte) error {
+	n, err := hex.Decode(dst, s)
+	if err != nil {
+		return &ParseError{"Invalid key: " + err.Error(), string(s)}
+	}
+	if n != wgkey.Size {
+		return &ParseError{"Keys must decode to exactly 32 bytes", string(s)}
+	}
+	return nil
+}
+
+func (cfg *Config) handleDeviceLine(key, value mem.RO, valueBytes []byte) error {
+	if key.EqualString("private_key") {
```

Comment:
This is much less readable than the switch statement. Is this performance critical?

Reply:
It's not possible to get a string here without allocating, so for memory it's pretty important. I can convert it into a switch statement, but it's not really better

Reply:
Right, but do these allocations matter in the bigger picture? If not, readable code is better.

Reply:
Allocating for this using a string copy does seem to be on the hot path and is actually a significant portion of the runtime, so I'm going to leave it as this.

I do understand that this is a lot less readable, but I think it's worth the cost.

Reply:
> Allocating for this using a string copy does seem to be on the hot path and is actually a significant portion of the runtime, so I'm going to leave it as this.

A significant portion of the runtime for what? For this benchmark specifically, or for something measurable in the client code? To put it another way, why does it matter that this benchmark be fast?

I'm all for making things fast in general. But when there's a significant readability regression, the need for the performance optimizations to matter gets more important.

As an example, the deephash optimizations (which had an awful bug--the dangers of optimization!) were important to keeping the hello server stable under load, and enabled deephash to be used in control.

Reply:
Mmm this is fair, and I probably don't have as much context on where `FromUAPI` is used as I should to know whether the benchmark I wrote is representative of the common use case in prod. That being said, it appears no matter what path the code takes through, it will hit one of the two switch statements which were changed into `EqualString`, and so however the API is used, as long as it's not empty (in which case I think the point of optimizing would've been moot in general), it will strictly increase the performance.

I hear you on the readability issue, but I personally lean towards always optimizing speed so I am a bit biased. Whether this optimization is _necessary_ is a good question.

To measure impact, would it be possible to look at pprof outputs from usage?

Reply:
> To measure impact, would it be possible to look at pprof outputs from usage?

I have spent a fair amount of time looking at pprof output from the client. I don't think that this code is at all important to performance right now, at least right now.

> I personally lean towards always optimizing speed so I am a bit biased.

I share that bias. But where performance really doesn't matter, and when that performance actively conflicts with readability, I prefer readability.